### PR TITLE
Fix set_ciphersuites ignore unknown ciphers.

### PR DESCRIPTION
**Pull request by OttoHollmann**
Fix SSL_CTX_set_ciphersuites() and SSL_set_ciphersuites() to ignore unknown cipehrs.
These functions now should behave like SSL_CTX_set_cipher_list() and SSL_set_cipher_list().

This commit fixes bug/feature https://github.com/openssl/openssl/issues/9335

### CODE REVIEW

**Comment by kaduk** [comment]
I had this vague recollection that the ciphersuites API was intentionally made to be strict and fail if something unknown was passed (maybe to avoid the risk of weird behavior when all components of the passed list were unknown?), but if Matt doesn't remember such a discussion I'm inclined to believe that I'm imagining things.

**Comment by OttoHollmann** [comment]
There is another option. Keep this ciphersuites API as is and allow TLS 1.3 ciphersuites to be entered in cipher_list. See my first PR https://github.com/openssl/openssl/pull/10590
Then users can decide which API they want to use.

**Comment by mattcaswell** [comment]
> but if Matt doesn't remember such a discussion I'm inclined to believe that I'm imagining things.

...or that I can't even remember what day of the week it is, let alone what previous discussions we might have had on this issue.

**Comment by mattcaswell** [comment]
> There is another option. Keep this ciphersuites API as is and allow TLS 1.3 ciphersuites to be entered in cipher_list. See my first PR #10590

I think this is problematic. The whole reason we didn't go down this route in the first place is that it is possible that legacy configurations have cipher lists which, if applied to the TLSv1.3 ciphersuites, would accidentally disable them. Given that TLSv1.3 ciphersuites are so different to TLSv1.2 ciphersuites we decided to separate them. You can argue the pros and cons of that decision - but I don't think we should be back-tracking on it now.

**Comment by OttoHollmann** [comment]
> I think this is problematic. The whole reason we didn't go down this route in the first place is that it is possible that legacy configurations have cipher lists which, if applied to the TLSv1.3 ciphersuites, would accidentally disable them. Given that TLSv1.3 ciphersuites are so different to TLSv1.2 ciphersuites we decided to separate them. You can argue the pros and cons of that decision - but I don't think we should be back-tracking on it now.

I know that this is very controversial theme, but I'm still think, that my improved cipher_list will deal with accidentally disabling ciphersuites. That's the reason why I added new parameters for specifying cipheruites version.
(Maybe we should move our further discussion into appropriate PR.)

File: ssl/ssl_ciph.c
Comment (mattcaswell) [inline ssl/ssl_ciph.c:1315]: Why is it that this instance returns 0, but the other "no cipher match" failure in this function a few lines above returns 0. Shouldn't that one also return 1?

File: ssl/ssl_ciph.c
Comment (mattcaswell) [inline ssl/ssl_ciph.c:1313]: Setting this error here no longer seems appropriate. Since we're going to continue to process the rest of the list we might get a subsequent cipher match. Probably we should delay raising this error until we notice that the returned stack has no elements in it in `set_ciphersuites`.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/ssl_ciph.c
Comment (OttoHollmann) [reply ssl/ssl_ciph.c:1313]: Yes, it should return 1.

File: ssl/ssl_ciph.c
Comment (OttoHollmann) [reply ssl/ssl_ciph.c]: Yes, error there is no longer appropriate. I've moved this error into function set_ciphersuites().

I was originally thinking about some "warnings", so user should know, that some of entered ciphersuites can't be recognized., that was reason why I keeped this error.

**Review by mattcaswell — APPROVED** [review]
Needs second review

**Comment by t8m** [comment]
Please note there is a merge commit in the PR. We do not allow merge commits so the merge commit needs to be dropped from the PR and the PR needs to be properly rebased.

**Comment by OttoHollmann** [comment]
Rebased.

File: ssl/ssl_ciph.c
Comment (t8m) [inline ssl/ssl_ciph.c:1333]: Nit: extra space after 0

**Review by t8m — CHANGES_REQUESTED** [review]
One trivial nit.

File: ssl/ssl_ciph.c
Comment (OttoHollmann) [reply ssl/ssl_ciph.c]: Removed

**Review by t8m — APPROVED** [review]
LGTM. @mattcaswell is it still approved?

**Review by mattcaswell — APPROVED** [review]
Still good.

**Comment by t8m** [comment]
I'm sorry we neglected this PR for so long time.

I've rebased the branch to fix conflicts. @mattcaswell can you please re-approve?

**Review by mattcaswell — APPROVED** [review]
Still good.

**Comment by t8m** [comment]
Merged to master. Unfortunately I forgot to squash the commits which I intended to do originally. Anyway thank you for your contribution and patience.

**Pull request closed without merging**
